File: src/main/scala/org/apache/datacommons/prepbuddy/rdds/TransformableRDD.scala
package org.apache.datacommons.prepbuddy.rdds

import org.apache.datacommons.prepbuddy.clusterers.{Cluster, ClusteringAlgorithm, Clusters, TextFacets}
import org.apache.datacommons.prepbuddy.imputations.ImputationStrategy
import org.apache.datacommons.prepbuddy.normalizers.NormalizationStrategy
import org.apache.datacommons.prepbuddy.smoothers.SmoothingMethod
import org.apache.datacommons.prepbuddy.types.{CSV, FileType}
import org.apache.datacommons.prepbuddy.utils.{PivotTable, RowRecord}
import org.apache.spark.rdd.RDD

class TransformableRDD(parent: RDD[String], fileType: FileType = CSV) extends AbstractRDD(parent, fileType) {

    private def getFileType: FileType = fileType

    /**
      * Returns a new RDD containing smoothed values of @columnIndex using @smoothingMethod
      *
      * @param columnIndex     Column Index
      * @param smoothingMethod Method that will be used for smoothing of the data
      * @return RDD<Double>
      */
    def smooth(columnIndex: Int, smoothingMethod: SmoothingMethod): RDD[Double] = {
        validateColumnIndex(columnIndex)
        validateNumericColumn(columnIndex)
        val cleanRDD: TransformableRDD = removeRows(!_.isNumberAt(columnIndex))
        val columnDataset: RDD[String] = cleanRDD.select(columnIndex)
        smoothingMethod.smooth(columnDataset)
    }

    /**
      * Returns a new TransformableRDD containing only the elements that satisfy the matchInDictionary.
      *
      * @param predicate A matchInDictionary function, which gives bool value for every row.
      * @return TransformableRDD
      */
    def removeRows(predicate: (RowRecord) => Boolean): TransformableRDD = {
        val filteredRDD = filter((record: String) => {
            val rowRecord = fileType.parse(record)
            !predicate(rowRecord)
        })
        new TransformableRDD(filteredRDD, fileType)
    }

    /**
      * Returns a new TransformableRDD by replacing the @cluster's text with specified @newValue
      *
      * @param cluster     Cluster of similar values to be replaced
      * @param newValue    Value that will be used to replace all the cluster value
      * @param columnIndex Column index
      * @return TransformableRDD
      */
    def replaceValues(cluster: Cluster, newValue: String, columnIndex: Int): TransformableRDD = {
        val mapped: RDD[String] = map((row) => {
            var rowRecord: RowRecord = fileType.parse(row)
            val value: String = rowRecord.select(columnIndex)
            if (cluster.containsValue(value)) rowRecord = rowRecord.replace(columnIndex, newValue)
            fileType.join(rowRecord)
        })
        new TransformableRDD(mapped, fileType)
    }

    /**
      * Returns a new TransformableRDD by applying the function on all rows marked as @flag
      *
      * @param symbol            Symbol that has been used for flagging.
      * @param symbolColumnIndex Symbol column index
      * @param mapFunction       map function
      * @return TransformableRDD
      */
    def mapByFlag(symbol: String, symbolColumnIndex: Int, mapFunction: (String) => String): TransformableRDD = {
        val mappedRDD: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val symbolColumn: String = rowRecord.select(symbolColumnIndex)
            if (symbolColumn.equals(symbol)) mapFunction(record) else record
        })
        new TransformableRDD(mappedRDD, fileType)
    }

    /**
      * Returns a new TransformableRDD that contains records flagged by @symbol
      * based on the evaluation of @markerPredicate
      *
      * @param symbol          Symbol that will be used to flag
      * @param markerPredicate A matchInDictionary which will determine whether to flag a row or not
      * @return TransformableRDD
      */
    def flag(symbol: String, markerPredicate: (RowRecord) => Boolean): TransformableRDD = {
        val flagged: RDD[String] = map((record) => {
            var newRow: String = fileType.appendDelimiter(record)
            val rowRecord: RowRecord = fileType.parse(record)
            if (markerPredicate(rowRecord)) newRow += symbol
            newRow
        })
        new TransformableRDD(flagged, fileType)
    }

    /**
      * Returns number of column in this rdd
      *
      * @return int
      */
    def numberOfColumns(): Int = columnLength

    /**
      * Returns Clusters that has all cluster of text of @columnIndex according to @algorithm
      *
      * @param columnIndex         Column Index
      * @param clusteringAlgorithm Algorithm to be used to form clusters
      * @return Clusters
      */
    def clusters(columnIndex: Int, clusteringAlgorithm: ClusteringAlgorithm): Clusters = {
        validateColumnIndex(columnIndex)
        val textFacets: TextFacets = listFacets(columnIndex)
        val rdd: RDD[(String, Int)] = textFacets.rdd
        val tuples: Array[(String, Int)] = rdd.collect

        clusteringAlgorithm.getClusters(tuples)
    }

    /**
      * Returns a new TextFacet containing the cardinal values of @columnIndex
      *
      * @param columnIndex index of the column
      * @return TextFacets
      */
    def listFacets(columnIndex: Int): TextFacets = {
        validateColumnIndex(columnIndex)
        val columnValuePair: RDD[(String, Int)] = map(record => (fileType.parse(record).select(columnIndex), 1))
        val facets: RDD[(String, Int)] = columnValuePair.reduceByKey(_ + _)
        new TextFacets(facets)
    }

    /**
      * Returns a RDD of double which is a product of the values in @firstColumn and @secondColumn
      *
      * @param firstColumn  First Column Index
      * @param secondColumn Second Column Index
      * @return RDD[Double]
      */
    def multiplyColumns(firstColumn: Int, secondColumn: Int): RDD[Double] = {
        validateColumnIndex(firstColumn :: secondColumn :: Nil)
        val rddOfNumbers: TransformableRDD = removeRows((record) => {
            !record.isNumberAt(firstColumn) || !record.isNumberAt(secondColumn)
        })
        rddOfNumbers.map((row) => {
            val firstColumnValue: String = fileType.parse(row).select(firstColumn)
            val secondColumnValue: String = fileType.parse(row).select(secondColumn)
            firstColumnValue.toDouble * secondColumnValue.toDouble
        })
    }

    /**
      * Generates a PivotTable by pivoting data in the pivotalColumn
      *
      * @param pivotalColumn            Pivotal Column
      * @param independentColumnIndexes Independent Column Indexes
      * @return PivotTable
      */
    def pivotByCount(pivotalColumn: Int, independentColumnIndexes: Seq[Int]): PivotTable[Integer] = {
        validateColumnIndex(independentColumnIndexes.+:(pivotalColumn).toList)
        val table: PivotTable[Integer] = new PivotTable[Integer](0)
        independentColumnIndexes.foreach((index) => {
            val facets: TextFacets = listFacets(pivotalColumn :: index :: Nil)
            val tuples = facets.rdd.collect()
            tuples.foreach((tuple) => {
                val split: Array[String] = tuple._1.split("\n")
                table.addEntry(split(0), split(1), tuple._2)
            })
        })
        table
    }

    /**
      * Returns a new TextFacet containing the facets of @columnIndexes
      *
      * @param columnIndexes List of column index
      * @return TextFacets
      */
    def listFacets(columnIndexes: List[Int]): TextFacets = {
        validateColumnIndex(columnIndexes)
        val columnValuePair: RDD[(String, Int)] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val joinedValue: String = rowRecord.select(columnIndexes).mkString("\n")
            (joinedValue, 1)
        })
        val facets: RDD[(String, Int)] = columnValuePair.reduceByKey(_ + _)
        new TextFacets(facets)
    }

    /**
      * Returns a TransformableRDD by splitting the @column according to the specified lengths
      *
      * @param column       Column index of the value to be split
      * @param fieldLengths List of integers specifying the number of character each split value will contains
      * @param retainColumn false when you want to remove the column value at @column in the result TransformableRDD
      * @return TransformableRDD
      */
    def splitByFieldLength(column: Int, fieldLengths: List[Int], retainColumn: Boolean = false): TransformableRDD = {
        validateColumnIndex(column)

        def splitString(value: String): Array[String] = {
            var result = Array.empty[String]
            var columnValue = value
            for (length <- fieldLengths) {
                val splitValue: String = {
                    if (columnValue.length >= length) columnValue.take(length) else columnValue
                }.mkString
                result = result.:+(splitValue)
                columnValue = columnValue.drop(length)
            }
            result
        }

        val transformed: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val splitValue: Array[String] = splitString(rowRecord.select(column))
            val result: RowRecord = arrangeRecords(rowRecord, column :: Nil, splitValue, retainColumn)
            fileType.join(result)
        })
        new TransformableRDD(transformed, fileType)
    }

    private def arrangeRecords(rowRecord: RowRecord, cols: List[Int], result: Array[String], retainColumn: Boolean) = {
        if (retainColumn) rowRecord.appendColumns(result) else rowRecord.valuesNotAt(cols).appendColumns(result)
    }

    /**
      * Returns a new TransformableRDD by splitting the @column by the delimiter provided
      *
      * @param column       Column index of the value to be split
      * @param delimiter    delimiter or regEx that will be used to split the value @column
      * @param retainColumn false when you want to remove the column value at @column in the result TransformableRDD
      * @param maxSplit     Maximum number of split to be added to the result TransformableRDD
      * @return TransformableRDD
      */
    def splitByDelimiter(column: Int, delimiter: String, retainColumn: Boolean = false, maxSplit: Int = -1):
    TransformableRDD = {
        validateColumnIndex(column)
        val transformed: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val splitValue: Array[String] = rowRecord.select(column).split(delimiter, maxSplit)
            val result: RowRecord = arrangeRecords(rowRecord, List(column), splitValue, retainColumn)
            fileType.join(result)
        })
        new TransformableRDD(transformed, fileType)
    }

    /**
      * Returns a new TransformableRDD by merging 2 or more columns together
      *
      * @param columns       List of columns to be merged
      * @param separator     Separator to be used to separate the merge value
      * @param retainColumns false when you want to remove the column value at @column in the result TransformableRDD
      * @return TransformableRDD
      */
    def mergeColumns(columns: List[Int], separator: String = " ", retainColumns: Boolean = false): TransformableRDD = {
        validateColumnIndex(columns)
        val transformedRDD: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val mergedValue: String = rowRecord.select(columns).mkString(separator)
            val result: RowRecord = arrangeRecords(rowRecord, columns, Array(mergedValue), retainColumns)
            fileType.join(result)
        })
        new TransformableRDD(transformedRDD, fileType)
    }

    /**
      * Returns a new TransformableRDD by normalizing values of the given column using different Normalizers
      *
      * @param columnIndex Column Index
      * @param normalizer  Normalization Strategy
      * @return TransformableRDD
      */
    def normalize(columnIndex: Int, normalizer: NormalizationStrategy): TransformableRDD = {
        validateColumnIndex(columnIndex)
        normalizer.prepare(this, columnIndex)
        val rdd: RDD[String] = map((record) => {
            val columns: RowRecord = fileType.parse(record)
            val normalizedColumn = normalizer.normalize(columns.select(columnIndex))
            val normalizedRecord: RowRecord = columns.replace(columnIndex, normalizedColumn)
            fileType.join(normalizedRecord)
        })
        new TransformableRDD(rdd, fileType)
    }

    /**
      * Returns a new TransformableRDD by imputing missing values of the @columnIndex using the @strategy
      *
      * @param column   Column index
      * @param strategy Imputation strategy
      * @return TransformableRDD
      */
    def impute(column: Int, strategy: ImputationStrategy): TransformableRDD = impute(column, strategy, List.empty)

    /**
      * Returns a new TransformableRDD by imputing missing values and @missingHints of the @columnIndex using the @strategy
      *
      * @param columnIndex  Column Index
      * @param strategy     Imputation Strategy
      * @param missingHints List of Strings that may mean empty
      * @return TransformableRDD
      */
    def impute(columnIndex: Int, strategy: ImputationStrategy, missingHints: List[String]): TransformableRDD = {
        validateColumnIndex(columnIndex)
        strategy.prepareSubstitute(this, columnIndex)
        val transformed: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val value: String = rowRecord.select(columnIndex)
            var replacementValue: String = value
            if (value == null || value.isEmpty || missingHints.contains(value)) {
                replacementValue = strategy.handleMissingData(rowRecord)
            }

            fileType.join(rowRecord.replace(columnIndex, replacementValue))
        })

        new TransformableRDD(transformed, fileType)
    }

    /**
      * Returns a new TransformableRDD by dropping the @columnIndex
      *
      * @param columnIndex The column that will be dropped.
      * @return TransformableRDD
      */
    def drop(columnIndex: Int, columnIndexes: Int*): TransformableRDD = {
        val columnsToBeDroped: List[Int] = columnIndexes.+:(columnIndex).toList
        validateColumnIndex(columnsToBeDroped)
        val transformed: RDD[String] = map((record: String) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val resultRecord: RowRecord = rowRecord.valuesNotAt(columnsToBeDroped)
            fileType.join(resultRecord)
        })
        new TransformableRDD(transformed, fileType)
    }

    /**
      * Returns a new RDD containing the duplicate values at the specified column
      *
      * @param columnIndex Column where to look for duplicates
      * @return RDD
      */
    def duplicatesAt(columnIndex: Int): RDD[String] = {
        validateColumnIndex(columnIndex)
        val specifiedColumnValues: RDD[String] = map(fileType.parse(_).select(columnIndex))
        new TransformableRDD(specifiedColumnValues, fileType).duplicates()
    }

    /**
      * Returns a new TransformableRDD containing duplicate records of this TransformableRDD by considering all the columns as primary key.
      *
      * @return TransformableRDD A new TransformableRDD consisting unique duplicate records.
      */
    def duplicates(): TransformableRDD = duplicates(List.empty)

    /**
      * Returns a new TransformableRDD containing unique duplicate records of this TransformableRDD by considering the given columns as primary key.
      *
      * @param primaryKeyColumns A list of integers specifying the columns that will be combined to create the primary key
      * @return TransformableRDD A new TransformableRDD consisting unique duplicate records.
      */
    def duplicates(primaryKeyColumns: List[Int]): TransformableRDD = {
        validateColumnIndex(primaryKeyColumns)
        val fingerprintedRecord: RDD[(Long, String)] = generateFingerprintedRDD(primaryKeyColumns)
        val recordsGroupedByKey: RDD[(Long, List[String])] = fingerprintedRecord.aggregateByKey(List.empty[String])(
            (accumulatorValues, currentValue) => accumulatorValues.::(currentValue),
            (aggregator1, aggregator2) => aggregator1 ::: aggregator2
        )
        val duplicateRecords: RDD[String] = recordsGroupedByKey.filter(_._2.size != 1).flatMap(_._2)
        new TransformableRDD(duplicateRecords, fileType).deduplicate()
    }

    /**
      * Returns a new TransformableRDD containing unique duplicate records of this TransformableRDD by considering all the columns as primary key.
      *
      * @return TransformableRDD A new TransformableRDD consisting unique duplicate records.
      */
    def deduplicate(): TransformableRDD = deduplicate(List.empty)

    /**
      * Returns a new TransformableRDD containing unique duplicate records of this TransformableRDD by considering the given columns as primary key.
      *
      * @param primaryKeyColumns A list of integers specifying the columns that will be combined to create the primary key
      * @return TransformableRDD A new TransformableRDD consisting unique duplicate records.
      */
    def deduplicate(primaryKeyColumns: List[Int]): TransformableRDD = {
        validateColumnIndex(primaryKeyColumns)
        val fingerprintedRDD: RDD[(Long, String)] = generateFingerprintedRDD(primaryKeyColumns)
        val reducedRDD: RDD[(Long, String)] = fingerprintedRDD.reduceByKey((accumulator, record) => record)
        new TransformableRDD(reducedRDD.values, fileType)
    }

    private def generateFingerprintedRDD(primaryKeyColumns: List[Int]): RDD[(Long, String)] = {
        map(record => {
            val rowRecord: RowRecord = fileType.parse(record)
            val fingerprint = rowRecord.fingerprintBy(primaryKeyColumns)
            (fingerprint, record)
        })
    }

    /**
      * Returns a new TransformableRDD containing the unique elements in the specified column
      *
      * @param columnIndex Column Index
      * @return RDD<String>
      */
    def unique(columnIndex: Int): RDD[String] = {
        validateColumnIndex(columnIndex)
        val specifiedColumnValues: RDD[String] = map(fileType.parse(_).select(columnIndex))
        new TransformableRDD(specifiedColumnValues, fileType).deduplicate()
    }

    /**
      * Zips the other TransformableRDD with this TransformableRDD and
      * returns a new TransformableRDD with current file format.
      * Both the TransformableRDD must have same number of records
      *
      * @param otherRDD Other TransformableRDD from where the columns will be added to this TransformableRDD
      * @return TransformableRDD
      */
    def addColumnsFrom(otherRDD: TransformableRDD): TransformableRDD = {
        val otherRDDInCurrentFileFormat = {
            if (this.getFileType != otherRDD.getFileType) {
                otherRDD.map(record => {
                    val rowRecord: RowRecord = otherRDD.getFileType.parse(record)
                    fileType.join(rowRecord)
                })
            }
            else {
                otherRDD
            }
        }
        val zippedRecords: RDD[(String, String)] = zip(otherRDDInCurrentFileFormat)
        val recordsWithAddedColumns: RDD[String] = zippedRecords.map(row => fileType.appendDelimiter(row._1) + row._2)

        new TransformableRDD(recordsWithAddedColumns, fileType)
    }

    /**
      * Returns a new TransformableRDD containing values of @columnIndexes
      *
      * @param columnIndexes A number of integer values specifying the columns that will be used to create the new table
      * @return TransformableRDD
      */
    def select(columnIndexes: List[Int]): TransformableRDD = {
        validateColumnIndex(columnIndexes)
        val selectedColumnValues: RDD[String] = map((record) => {
            val rowRecord: RowRecord = fileType.parse(record)
            val resultValues: RowRecord = rowRecord.select(columnIndexes)
            fileType.join(resultValues)
        })
        new TransformableRDD(selectedColumnValues, fileType)
    }
}

File: python/pyprepbuddy/rdds/transformable_rdd.py
from py4j.java_gateway import java_import
from py4j.protocol import Py4JJavaError
from pyspark import RDD, StorageLevel
from pyspark.mllib.common import _java2py

from pyprepbuddy import py2java_int_list
from pyprepbuddy.class_names import ClassNames
from pyprepbuddy.cluster.clusters import Clusters
from pyprepbuddy.cluster.text_facets import TextFacets
from pyprepbuddy.exceptions.application_exception import ApplicationException
from pyprepbuddy.serializer import BuddySerializer
from pyprepbuddy.utils.pivot_table import PivotTable


class TransformableRDD(RDD):
    """
    A Transformable Resilient Distributed Dataset is extended from RDD, the basic abstraction in Spark.
    It represents an immutable, partitioned collection of elements that can be operated on in parallel.
    It provides a bunch of higher level functionality over on RDD.
    """
    def __init__(self, rdd, file_type='CSV', t_rdd=None, sc=None):
        if rdd is not None:
            jvm = rdd.ctx._jvm
            java_import(jvm, ClassNames.BYTES_TO_STRING)
            java_import(jvm, ClassNames.TRANSFORMABLE_RDD)

            self.__set_file_type(jvm, file_type)
            self.spark_context = rdd.ctx
            java_rdd = rdd._reserialize(BuddySerializer())._jrdd.map(jvm.BytesToString())
            self._transformable_rdd = jvm.JavaTransformableRDD(java_rdd, self.__file_type)
            RDD.__init__(self, rdd._jrdd, rdd.ctx)
        else:
            jvm = sc._jvm
            java_import(jvm, ClassNames.STRING_TO_BYTES)
            self.spark_context = sc
            self.__set_file_type(jvm, file_type)
            self._transformable_rdd = t_rdd
            rdd = t_rdd.map(jvm.StringToBytes())
            RDD.__init__(self, rdd, sc, BuddySerializer())

    def __set_file_type(self, jvm, file_type):
        java_import(jvm, ClassNames.FileType)
        file_types = {
            'CSV': jvm.FileType.CSV,
            'TSV': jvm.FileType.TSV
        }
        if file_type in file_types.values():
            self.__file_type = file_type
        elif file_type.upper() in file_types:
            self.__file_type = file_types[file_type.upper()]
        else:
            raise ValueError('"%s" is not a valid file type\nValid file types are CSV and TSV' % file_type)

    def deduplicate(self, column_indexes=None):
        """
        Returns a new TransformableRDD containing only unique records by considering all the columns as the primary key.
        :param column_indexes: Sequence of the column index
        :return: TransformableRDD
        """
        if column_indexes is None:
            return TransformableRDD(None, self.__file_type, self._transformable_rdd.deduplicate(),
                                    sc=self.spark_context)
        return TransformableRDD(None, self.__file_type, self._transformable_rdd.deduplicate(column_indexes),
                                sc=self.spark_context)

    def impute(self, column_index, imputation_strategy):
        """
        Returns a new TransformableRDD by imputing missing values of the @column_index using the @imputation_strategy
        :param column_index: index of the column on which imputation will be performed
        :param imputation_strategy: strategy for impute the missing value
        :return: TransformableRDD
        """
        strategy_apply = imputation_strategy._get_strategy(self.spark_context)
        return TransformableRDD(None,
                                self.__file_type,
                                self._transformable_rdd.impute(column_index, strategy_apply),
                                sc=self.spark_context)

    def clusters(self, column_index, clustering_algorithm):
        """
        Returns Clusters that has all cluster of text of @column_index according to @clustering_algorithm
        :param column_index: index of the column
        :param clustering_algorithm: Algorithm to be used to form clusters
        :return: Clusters
        """
        algorithm = clustering_algorithm._get_algorithm(self.spark_context)
        return Clusters(self._transformable_rdd.clusters(column_index, algorithm))

    def list_facets_of(self, column_index):
        """
        Returns a new TextFacet containing the facets of @columnIndex
        :param column_index: Index of the column
        :return: TextFacets
        """
        try:
            return TextFacets(self._transformable_rdd.listFacets(column_index))
        except Py4JJavaError as e:
            java_exception = e.java_exception
            raise ApplicationException(java_exception)

    def list_facets(self, column_indexes):
        """
        Returns a new TextFacet containing the facets of @column_indexes
        :param column_indexes: Sequence of column index
        :return: TextFacets
        """
        array = py2java_int_list(self.spark_context, column_indexes)
        return TextFacets(self._transformable_rdd.listFacets(array))

    def select(self, column_index, *column_indexes):
        """
        Returns RDD of given column and TransformableRDD for multiple columns.
        :param column_index: index of the one column
        :param column_indexes: Sequence of column index
        :returns: RDD,TransformableRDD
        """
        if column_indexes.__len__() == 0:
            return self._transformable_rdd.select(column_index)
        java_array = py2java_int_list(self.spark_context, column_indexes)
        java_array.add(0, column_index)
        java_rdd = self._transformable_rdd.select(java_array)
        return TransformableRDD(None, self.__file_type, java_rdd, sc=self.spark_context)

    def normalize(self, column_index, normalizer_strategy):
        """
        Returns a new TransformableRDD by normalizing values of the given column using different Normalizers
        :param column_index: Index of the column
        :param normalizer_strategy: normalization strategy by which you want to normalize
        :return: TransformableRDD
        """
        normalizer = normalizer_strategy._get_normalizer(self.spark_context)
        return TransformableRDD(None, self.__file_type, self._transformable_rdd.normalize(column_index, normalizer),
                                sc=self.spark_context)

    def smooth(self, column_index, smoothing_method):
        """
        Returns a new RDD containing smoothed values of @column_index using @smoothing_method
        :param column_index: Index of the column
        :param smoothing_method: smoothing method by which you want to smooth data
        :return: RDD
        """
        method = smoothing_method._get_smoothing_method(self.spark_context)
        rdd = self._transformable_rdd.smooth(column_index, method)
        return _java2py(self.spark_context, rdd.rdd())

    def merge_columns(self, list_of_columns, separator=" ", retain_columns=False):
        """
        Returns a new TransformableRDD containing the merged column of given indexes with the @separator
        :param list_of_columns:Sequence of column index
        :param separator:the delimiter string reference by which merging will be performed
        :param retain_columns:preserves the column if true
        :return: TransformableRDD
        """
        int_list = py2java_int_list(self.spark_context, list_of_columns)
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.mergeColumns(int_list, separator, retain_columns),
                                sc=self.spark_context)

    def split_by_delimiter(self, column_index, delimiter, retain_column=False):
        """
        Returns a new TransformableRDD containing split columns by specified @delimiter on the given index
        :param column_index: index of the column on which operation will be performed
        :param delimiter:the delimiter string reference by which split will be performed
        :param retain_column:preserves the column if true
        :return: TransformableRDD
        """
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.splitByDelimiter(column_index, delimiter, retain_column),
                                sc=self.spark_context)

    def split_by_field_length(self, column_index, field_lengths, retain_column=False):
        """
        Returns a new TransformableRDD containing split columns using @split_plan
        :param column_index:index of the column on which operation will be performed
        :param field_lengths:Sequence of lengths of the resultant column used for splitting the element on @column_index
        :param retain_column:preserves the column if true
        :return: TransformableRDD
        """
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.splitByFieldLength(column_index, field_lengths, retain_column),
                                sc=self.spark_context)

    def get_duplicates(self, column_indexes=None):
        """
        Returns a new TransformableRDD containing unique duplicate records of this TransformableRDD by considering the
        given columns as primary key.
        :param column_indexes: Sequence of column indexes
        :return: TransformableRDD
        """
        if column_indexes is None:
            return TransformableRDD(None, self.__file_type,
                                    self._transformable_rdd.duplicates(),
                                    sc=self.spark_context)
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.duplicates(column_indexes),
                                sc=self.spark_context)

    def drop_column(self, column_index):
        """
        Returns a new TransformableRDD by dropping the column at given index
        :param column_index: Index of the column which will be dropped.
        :return:
        """
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.drop(column_index),
                                sc=self.spark_context)

    def replace_values(self, one_cluster, new_value, column_index):
        """
        Returns a new TransformableRDD by replacing the @cluster's text with specified @new_value
        :param one_cluster: Cluster of similar values to be replaced
        :param new_value: Value that will be used to replace all the cluster value
        :param column_index: Index of the column which will be replaced.
        :return: TransformableRDD
        """
        cluster = one_cluster.get_cluster()
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.replaceValues(cluster, new_value, column_index),
                                sc=self.spark_context)

    def multiply_columns(self, first_column, second_column):
        """
        Returns a RDD which is a product of the values in @first_column and @second_column
        :param first_column: One column index
        :param second_column: Another column index
        :return: RDD
        """
        _rdd = self._transformable_rdd.multiplyColumns(first_column, second_column).rdd()
        return _java2py(self.spark_context, _rdd)

    def to_double_rdd(self, column_index):
        """
        Returns a RDD by converting values to double of given column index
        :param column_index:One column index in TransformableRDD
        :return:RDD
        """
        rdd = self._transformable_rdd.toDoubleRDD(column_index).rdd()
        return _java2py(self.spark_context, rdd)

    def add_columns_from(self, other):
        return TransformableRDD(None, self.__file_type,
                                self._transformable_rdd.addColumnsFrom(other._transformable_rdd),
                                sc=self.spark_context)

    def pivot_by_count(self, pivotal_column, independent_column_indexes):
        """
        Generates a PivotTable by pivoting data in the pivotalColumn
        :param pivotal_column: Pivotal column index
        :param independent_column_indexes: Independent column indexes
        :return: PivotTable
        """
        column_indexes = py2java_int_list(self.spark_context, independent_column_indexes)
        return PivotTable(self._transformable_rdd.pivotByCount(pivotal_column, column_indexes))

    def map(self, function, preservesPartitioning=False):
        return TransformableRDD(super(TransformableRDD, self).map(function, preservesPartitioning), self.__file_type)

    def filter(self, f):
        return TransformableRDD(super(TransformableRDD, self).filter(f), self.__file_type)

    def cache(self):
        return TransformableRDD(super(TransformableRDD, self).cache(), self.__file_type)

    def coalesce(self, num_partitions, shuffle=False):
        return TransformableRDD(super(TransformableRDD, self).coalesce(num_partitions, shuffle), self.__file_type)

    def distinct(self, num_partitions=None):
        return TransformableRDD(super(TransformableRDD, self).distinct(num_partitions), self.__file_type)

    def flatMap(self, func, preserves_partitioning=False):
        return TransformableRDD(super(TransformableRDD, self).flatMap(func, preserves_partitioning), self.__file_type)

    def intersection(self, other):
        return TransformableRDD(super(TransformableRDD, self).intersection(other), self.__file_type)

    def persist(self, storage_level=StorageLevel.MEMORY_ONLY_SER):
        return TransformableRDD(super(TransformableRDD, self).persist(storage_level), self.__file_type)

    def unpersist(self):
        return TransformableRDD(super(TransformableRDD, self).unpersist(), self.__file_type)

    def union(self, other):
        return TransformableRDD(super(TransformableRDD, self).union(other), self.__file_type)

    def mapPartitions(self, func, preserves_partitioning=False):
        return TransformableRDD(super(TransformableRDD, self).mapPartitions(func, preserves_partitioning),
                                self.__file_type)

    def sortBy(self, keyfunc, ascending=True, num_partitions=None):
        return TransformableRDD(super(TransformableRDD, self).sortBy(keyfunc, ascending, num_partitions),
                                self.__file_type)

    def setName(self, name):
        return TransformableRDD(super(TransformableRDD, self).setName(name), self.__file_type)

    def subtract(self, other, num_partitions=None):
        return TransformableRDD(super(TransformableRDD, self).subtract(other, num_partitions), self.__file_type)

    def subtractByKey(self, other, num_partitions=None):
        return TransformableRDD(super(TransformableRDD, self).subtractByKey(other, num_partitions), self.__file_type)

    def sample(self, with_replacement, fraction, seed=None):
        return TransformableRDD(super(TransformableRDD, self).sample(with_replacement, fraction, seed),
                                self.__file_type)

    def repartition(self, num_partitions):
        return TransformableRDD(super(TransformableRDD, self).repartition(num_partitions), self.__file_type)

    def pipe(self, command, env=None, check_code=False):
        return TransformableRDD(super(TransformableRDD, self).pipe(command, env, check_code), self.__file_type)
